### PR TITLE
Feature/Factor Graph BP: Add virtual evidence

### PR DESCRIPTION
**Pull request by tristantreb**
This PR adds capacity to handle virtual evidence on the factor graph BP.
I defined virtual evidence as a list of tabular CPD to be consistent with the current implementation

### Your checklist for this pull request

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### List of changes to the codebase in this pull request
- Add ability to support virtual evidence Factor Graph BP.
- In the base Inference class, exported the virtual evidence checks from `_virtual_evidence()` into a new function `_check_virtual_evidence()`. Indeed in the Factor Graph BP, there's no need to add a new variable to the graph, instead each evidence becomes a virtual message that gets added to the variable's incoming messages list
- Exhaustive testing
- Improved documentation

### CODE REVIEW

File: pgmpy/inference/ExactInference.py [hunk 1]
```diff
@@ -1274,8 +1274,14 @@ def query(self, variables, evidence):
...
             list of variables for which you want to compute the posterior
-        evidence: dict
+        evidence: dict (default: {})
             a dict key, value pair as {var: state_of_var_observed}
+        virtual_evidence: dict (default: {})
+            a dict key, virtual message list pair as {var: [virtual_message_array]}
+            Note: in message-parsing BP, handling virtual evidence is more straightforward thatn in the juction
+            tree BP implemented in BeliefPropagation. Each virtual evidence becomes a virtual message that gets
+            added to the list of computed messages incoming to the variable node. Hence, a dict makes the code
+            faster/cleaner.
```
Comment (tristantreb) [inline pgmpy/inference/ExactInference.py:1284]: Comment on choice for virtual evidence implementation

File: pgmpy/inference/ExactInference.py [hunk 2]
```diff
@@ -1274,8 +1274,14 @@ def query(self, variables, evidence):
...
             list of variables for which you want to compute the posterior
-        evidence: dict
+        evidence: dict (default: {})
             a dict key, value pair as {var: state_of_var_observed}
+        virtual_evidence: dict (default: {})
+            a dict key, virtual message list pair as {var: [virtual_message_array]}
+            Note: in message-parsing BP, handling virtual evidence is more straightforward that in the junction
+            tree BP implemented in the `BeliefPropagation` class. Each virtual evidence becomes a virtual message 
+            that gets added to the list of computed messages incoming to the variable node. Hence, a dict makes the 
+            code faster/cleaner than a list of TabularCPD.
```
Comment (tristantreb) [inline pgmpy/inference/ExactInference.py:1284]: Comment on choice for virtual evidence implementation

Comment (tristantreb) [inline pgmpy/inference/ExactInference.py:1284]: Comment on choice for virtual evidence implementation

File: pgmpy/inference/ExactInference.py [hunk 3]
```diff
@@ -1274,8 +1274,15 @@ def query(self, variables, evidence):
...
-        evidence: dict
+        evidence: dict (default: {})
             a dict key, value pair as {var: state_of_var_observed}
+        virtual_evidence: dict (default: {})
+            a dict key, virtual message list pair as {var: [virtual_message_array]}
+            Note: in message-parsing BP, handling virtual evidence is more straightforward that in the junction
+            tree BP implemented in the `BeliefPropagation` class. Each virtual evidence becomes a virtual message
+            that gets added to the list of computed messages incoming to the variable node. There's no need to use
+            the `Inference._virtual_evidence()` method. Hence, a dict with arrays makes the code faster/cleaner 
+            than a list of TabularCPD.
```
Comment (tristantreb) [inline pgmpy/inference/ExactInference.py:1285]: Comment on choice for virtual evidence implementation. Another option to be consistent with the main BeliefPropagation class is to let `query` take in `virtual_evidence` as a list of tabular CPD, transform it into a dict of var, np.array which is easier to parse, and then propagate the `virtual_evidence` as a dict into the code. I went to for the lighter option. Let me know what you prefer.

File: pgmpy/inference/ExactInference.py [hunk 4]
```diff
@@ -1274,8 +1274,15 @@ def query(self, variables, evidence):
...
-        evidence: dict
+        evidence: dict (default: {})
             a dict key, value pair as {var: state_of_var_observed}
+        virtual_evidence: dict (default: {})
+            a dict key, virtual message list pair as {var: [virtual_message_array]}
+            Note: in message-parsing BP, handling virtual evidence is more straightforward that in the junction
+            tree BP implemented in the `BeliefPropagation` class. Each virtual evidence becomes a virtual message
+            that gets added to the list of computed messages incoming to the variable node. There's no need to use
+            the `Inference._virtual_evidence()` method. Hence, a dict with arrays makes the code faster/cleaner
+            than a list of TabularCPD.
```
Comment (tristantreb) [inline pgmpy/inference/ExactInference.py:1285]: Comment on choice for virtual evidence implementation. Another option to be consistent with the main BeliefPropagation class is to let `query` take in `virtual_evidence` as a list of tabular CPD, transform it into a dict of var, np.array which is easier to parse, and then propagate the `virtual_evidence` as a dict into the code. I went to for the lighter option. Let me know what you prefer.

**Comment by ankurankan** [comment]
@tristantreb Thanks a lot for the PR. Would it be possible to add support for Bayesian Networks and Markov Networks in the `BeliefPropagationWithMessageParsing`? Essentially converting both of them to factor graphs and running the algorithm on that. It would give us a way to compare the inference results among all inference algorithms and make sure that all of them match.

**Pull request merged**